File: python/day11.py
from intcode import *
from collections import defaultdict
from helpers import cat


def paint(instr, initial):
    painted = defaultdict(int)
    pos = 0j
    painted[pos] = initial
    d = -1j
    turns = [-1j, 1j]
    comp = Computer(instr)
    while comp.state != State.Halted:
        color = comp.receive_run_output(painted[pos])
        turn = turns[comp.get_from_out_queue()]
        painted[pos] = color
        d *= turn
        pos += d
    return painted

def part1(instr):
    return len(paint(instr, 0))

def part2(instr):
    painted = paint(instr, 1)
    registration = [40 * [' '] for _ in range(6)]
    for p in painted:
        if painted[p]:
            registration[int(p.imag)][int(p.real)] = '#'
    return registration

def pretty_print(registration):
    for row in registration:
        print(cat(row))


instr = read_intcode_input(11)

print(part1(instr))
pretty_print(part2(instr))

File: clojure/day11.clj
(ns day11
  (:require [intcode :as ic]
            [aoc-utils.core :as aoc]))


(defn- paint [computer initial]
  (let [turns [aoc/left-turn aoc/right-turn]]
    (loop [comp computer
           pos [0 0]
           dir [0 -1]
           painted {pos initial}]
      (if (= :halted (:status comp))
        painted
        (let [comp' (ic/in-run-out comp (painted pos 0))
              turn (:output comp')
              comp'' (ic/pop-out-queue comp')
              color (:output comp'')
              dir' ((turns turn) dir)]
          (recur comp''
                 (aoc/pt+ pos dir')
                 dir'
                 (assoc painted pos color)))))))


(defn solve [filename]
  (let [computer (ic/initialize-from-file filename)]
    [(count (paint computer 0))
     (->> (paint computer 1)
          (keep (fn [[pt v]] (when (not (zero? v)) pt)))
          set
          aoc/show-grid
          #_println)]))


(solve 11)
